overlord/hookstate/hookmgr.go

```diff
@@ -296,16 +332,64 @@ func runHookAndWait(snapName string, revision snap.Revision, hookName, hookConte
 		close(hookCompleted)
 	}()
 
-	select {
-	// Hook completed; it may or may not have been successful.
-	case <-hookCompleted:
-		return buffer.Bytes(), hookError
+	var abortOrTimeoutError error
+	dyingCh := tomb.Dying()
+out:
+	for {
+		select {
+		// Hook completed; it may or may not have been successful.
+		case <-hookCompleted:
+			break out
+		// Max timeout reached, process will get killed below
+		case <-killTimerCh:
+			abortOrTimeoutError = fmt.Errorf("exceeded maximum runtime of %s", maxRuntime)
+		// Hook was aborted, process will get killed below
+		case <-dyingCh:
+			dyingCh = nil
+			abortOrTimeoutError = fmt.Errorf("hook %q aborted", hookName)
+		// cmdWaitTimeout was reached, i.e. command.Wait() did not
+		// finish in a reasonable amount of time, we can not use
+		// buffer in this case so return without it.
+		case <-cmdWaitTimerCh:
+			return nil, fmt.Errorf("%v, but did not stop", abortOrTimeoutError)
+		}
 
-	// Hook was aborted.
-	case <-tomb.Dying():
-		if err := command.Process.Kill(); err != nil {
+		// select above exited which means that aborted or killTimeout
+		// was reached. Kill the command and wait for command.Wait()
+		// to clean it up (but limit the wait with the cmdWaitTimer)
+		if cmdWaitTimerCh != nil {
+			continue
+		}
+		if err := killemAll(command); err != nil {
 			return nil, fmt.Errorf("cannot abort hook %q: %s", hookName, err)
 		}
-		return nil, fmt.Errorf("hook %q aborted", hookName)
+		cmdWaitTimer = time.NewTimer(cmdWaitTimeout)
+		defer cmdWaitTimer.Stop()
+		cmdWaitTimerCh = cmdWaitTimer.C
+		killTimerCh = nil
+	}
+
+	if abortOrTimeoutError != nil {
+		return buffer.Bytes(), abortOrTimeoutError
+	}
+	return buffer.Bytes(), hookError
+}
+
+var errtrackerReport = errtracker.Report
+
+func reportHookFailureOnErrtracker(context *Context, err error) {
+	errmsg := err.Error()
+	dupSig := fmt.Sprintf("%s:%s:%s", context.SnapName(), context.HookName(), err)
+	extra := map[string]string{
+		"HookName": context.HookName(),
+	}
+	if context.setup.IgnoreFail {
+		extra["IgnoreFail"] = "1"
+	}
+	oopsid, err := errtrackerReport(context.SnapName(), errmsg, dupSig, extra)
```

Comment:
Errmsg was a bit too terse, I changed that now to something more human readable. If you prefer it to be more symmetrical to dupSig, I can do that as well.